host/ecallids.c

```diff
@@ -0,0 +1,206 @@
+                atexit(_free_ecall_table);
+            }
+            else
+            {
+                _ecall_table_capacity *= 2;
+            }
+            _ecall_table = oe_realloc(
+                (void*)_ecall_table, _ecall_table_capacity * sizeof(char*));
+            if (!_ecall_table)
+                OE_RAISE(OE_OUT_OF_MEMORY);
+        }
+
+        _ecall_table[_ecall_table_size] = name;
+        *global_id = _ecall_table_size;
+        _ecall_table_size++;
+    }
+
+    result = OE_OK;
+done:
+    return result;
+}
+
+oe_result_t oe_get_global_id(const char* name, uint64_t* global_id)
+{
+    oe_result_t result = OE_UNEXPECTED;
+    bool locked = false;
+    if (oe_mutex_lock(&_lock) != 0)
+        OE_RAISE(OE_FAILURE);
+    locked = true;
+
+    result = _get_global_id(name, global_id);
+done:
+    if (locked)
+        oe_mutex_unlock(&_lock);
+    return result;
+}
+
+oe_result_t oe_get_ecall_ids(
+    oe_enclave_t* enclave,
+    const char* name,
+    uint64_t* global_id,
+    uint64_t* id)
+{
+    oe_result_t result = OE_FAILURE;
+    oe_ecall_id_t* ecall_id_table = NULL;
+    uint64_t ecall_id_table_size = 0;
+
+    /* Validate parameters */
+    if (!enclave || !global_id || !name || !id)
+    {
+        result = OE_INVALID_PARAMETER;
+        goto done;
+    }
+
+    /* Initialize output parameter to NULL id */
+    *id = OE_ECALL_ID_NULL;
+
+    /* Fetch ecall_id_table using platform agnostic API */
+    OE_CHECK(
+        oe_get_ecall_id_table(enclave, &ecall_id_table, &ecall_id_table_size));
+
+    /* Lookup/assign global id */
+    if (*global_id == OE_GLOBAL_ECALL_ID_NULL)
+    {
+        /* The global_id variable is generated by edger8r, one global id
+         *  variable per ecall. The same variable will be reused every time that
+         * ecall is made. Therefore, for an ecall, the following code will be
+         * executed only once.
+         */
+        OE_CHECK(oe_get_global_id(name, global_id));
+    }
+
+    /* Look up function id based on global id. */
+    if (*global_id >= ecall_id_table_size)
+        OE_RAISE(OE_NOT_FOUND);
+
+    /* Look-up the ecall id from the per-enclave table.
+     * The ecall_id_table is expected to be set up during enclave creation
+     */
+    *id = ecall_id_table[*global_id].id;
+    if (*id == OE_ECALL_ID_NULL)
+        OE_RAISE(OE_NOT_FOUND);
+
+    result = OE_OK;
+done:
+    return result;
+}
+
+oe_result_t oe_register_ecalls(
+    oe_enclave_t* enclave,
+    const oe_ecall_info_t* ecall_info_table,
+    uint32_t num_ecalls)
+{
+    oe_result_t result = OE_UNEXPECTED;
+    oe_ecall_id_t* ecall_id_table = NULL;
+    uint64_t max_global_id = 0;
+    uint64_t ecall_id_table_size = 0;
+    bool locked = false;
+
+    /* Validate parameters */
+    if (!enclave || !ecall_info_table || !num_ecalls)
+        OE_RAISE(OE_INVALID_PARAMETER);
+
+    if (oe_mutex_lock(&_lock) != 0)
+        OE_RAISE(OE_FAILURE);
```

Comment:
Avoid using the uninformative `OE_FAILURE` where possible. `oe_mutex_lock` returns an `oe_result_t` so we should return the underlying error here. i.e. you could `OE_CHECK(oe_mutex_lock(&_lock));` here.

Also in `oe_get_global_id`

Reply:
The host side threading APIs return int.
https://github.com/openenclave/openenclave/blob/7060b014b6a7af86eb6329ca49e07296dda9fe3e/host/hostthread.h#L143

Created issue for the inconsistency : #3486